### ISSUE TITLE
Better exception handling when env vars are missing for flask CLI

### ISSUE THREAD

**Issue by rochacbruno** [pallets/flask#2741]
### Expected Behavior

As there is the support for `lazy loading` the app, when running `flask` CLI without providing the proper environment variables we must see a better warning instead of raw exception.

Tell us what should happen.

**we should see a better warning or message pointing to the problem**

```pytb
# there is no FLASK_APP env var
$ flask --help
WARNING: You need to define the app e.g: `export FLASK_APP=app.py`
```
### Actual Behavior

Tell us what happens instead.

**we see traceback before the help message**

```pytb
# there is no FLASK_APP env var
$ flask --help                                                               Sat 28 Apr 2018 01:25:50 PM -03
Traceback (most recent call last):
  File "~Projects/personal/flasgger/venv/lib/python3.6/site-packages/flask/cli.py", line 235, in locate_app
    __import__(module_name)
ModuleNotFoundError: No module named 'app'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/Projects/personal/flasgger/venv/lib/python3.6/site-packages/flask/cli.py", line 529, in list_commands
    rv.update(info.load_app().cli.list_commands(ctx))
  File "~/Projects/personal/flasgger/venv/lib/python3.6/site-packages/flask/cli.py", line 372, in load_app
    app = locate_app(self, import_name, name)
  File "~/Projects/personal/flasgger/venv/lib/python3.6/site-packages/flask/cli.py", line 246, in locate_app
    'Could not import "{name}".'.format(name=module_name)
flask.cli.NoAppException: Could not import "app".
Usage: flask [OPTIONS] COMMAND [ARGS]...

  A general utility script for Flask applications.

  Provides commands from Flask, extensions, and the application. Loads the
  application defined in the FLASK_APP environment variable, or from a
  wsgi.py file. Setting the FLASK_ENV environment variable to 'development'
  will enable debug mode.

    $ export FLASK_APP=hello.py
    $ export FLASK_ENV=development
    $ flask run

Options:
  --version  Show the flask version
  --help     Show this message and exit.

Commands:
  routes  Show the routes for the app.
  run     Runs a development server.
  shell   Runs a shell in the app context.

```

The same happens to `run`

```pytb
$ flask run                                                          429ms  Sat 28 Apr 2018 01:32:48 PM -03
 * Serving Flask app "app.py"
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
Usage: flask run [OPTIONS]

Error: Could not import "app".
```

The `Error: Could not import "app".` could include `WARNING: You need to define the app e.g: export FLASK_APP=app.py`

### Suggestion

We could check the existence of `FLASK_APP` envvar before running any of the commands in the Group Cli, if FLASK_APP does not exist the dispatch of commands never happens.

### Environment

* Python version: 3.6.0
* Flask version: Flask==1.0
* Werkzeug version: Werkzeug==0.14.1
* Click: click==6.7

**Comment by rochacbruno** [comment]
* I can implement that and send a PR, just want to know if the idea is acceptable

**Comment by davidism** [comment]
Anything that improves help messages is great.

**Comment by miguelgrinberg** [comment]
This issue will be automatically closed due to being inactive for more than six months. Seeing that I haven't responded to your last comment, it is quite possible that I have dropped the ball on this issue and I apologize about that. If that is the case, do not take the closing of the issue personally as it is an automated process doing it, just reopen it and I'll get back to you.

**Comment by nerixim** [comment]
Looks better in flask 1.0.2, so maybe we can close this?

- with a `help` option

```
~/flask flask --help
Traceback (most recent call last):
  File "/Users/username/.local/share/virtualenvs/flask-C1e3YqV7/lib/python3.7/site-packages/flask/cli.py", line 529, in list_commands
    rv.update(info.load_app().cli.list_commands(ctx))
  File "/Users/username/.local/share/virtualenvs/flask-C1e3YqV7/lib/python3.7/site-packages/flask/cli.py", line 384, in load_app
    'Could not locate a Flask application. You did not provide '
flask.cli.NoAppException: Could not locate a Flask application. You did not provide the "FLASK_APP" environment variable, and a "wsgi.py" or "app.py" module was not found in the current directory.
Usage: flask [OPTIONS] COMMAND [ARGS]...

  A general utility script for Flask applications.

  Provides commands from Flask, extensions, and the application. Loads the
  application defined in the FLASK_APP environment variable, or from a
  wsgi.py file. Setting the FLASK_ENV environment variable to 'development'
  will enable debug mode.

    $ export FLASK_APP=hello.py
    $ export FLASK_ENV=development
    $ flask run

Options:
  --version  Show the flask version
  --help     Show this message and exit.

Commands:
  routes  Show the routes for the app.
  run     Runs a development server.
  shell   Runs a shell in the app context.
```

- without any options:

```
 ~/flask flask run   
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: off
Usage: flask run [OPTIONS]

Error: Could not locate a Flask application. You did not provide the "FLASK_APP" environment variable, and a "wsgi.py" or "app.py" module was not found in the current directory.
```

**Comment by Cgspc** [comment]
Hola que tal.
Estoy obteniendo el siguiente mensaje al efectuar flask run
* Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
Traceback (most recent call last):
  File "/home/claudio/python/flask_env/bin/flask", line 8, in <module>
    sys.exit(main())
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 966, in main
    cli.main(prog_name="python -m flask" if as_module else None)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 586, in main
    return super(FlaskGroup, self).main(*args, **kwargs)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 848, in run_command
    app = DispatchingApp(info.load_app, use_eager_loading=eager_loading)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 305, in __init__
    self._load_unlocked()
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 330, in _load_unlocked
    self._app = rv = self.loader()
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 392, in load_app
    app = locate_app(self, import_name, None, raise_if_not_found=False)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 257, in locate_app
    return find_best_app(script_info, module)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 83, in find_best_app
    app = call_factory(script_info, app_factory)
  File "/home/claudio/python/flask_env/lib/python3.7/site-packages/flask/cli.py", line 117, in call_factory
    return app_factory(script_info)
  File "/home/claudio/python/servpc_spa/app/__init__.py", line 20, in create_app
    app.config.from_object(app_config[config_name])
KeyError: <flask.cli.ScriptInfo object at 0x7fd50d307e48>

**Comment by ThiefMaster** [comment]
what made you think we speak Spanish?

**Comment by ilyes-23** [comment]
I have this problem too when I run flask app I wonder it what the solution

**Comment by davidism** [comment]
The various exception messages, including when the env var wasn't set and a default wasn't found, have improved over time. That one now reads:

> Could not locate a Flask application. You did not provide the "FLASK_APP" environment variable, and a "wsgi.py" or "app.py" module was not found in the current directory.

Tracebacks were added in #2208 to avoid confusion caused by the previous behavior where `list_commands()` hid errors completely.

After reviewing the code, I think we should show only the error message for `NoAppException`, while still showing the full traceback for other exceptions. Tracebacks for loading errors are no longer helpful as they only show Flask internals. We still want to show the help text (or other output) because the `flask` command itself isn't what's failing. I've also noticed that `get_command` still completely hides the loading error, resulting in only "no command named X", so that can be improved too.
